### ISSUE TITLE
Running on Rosetta on Apple Silicon (via Blender)

### ISSUE THREAD

**Issue by EclipsedButter** [pleiszenburg/zugbruecke#103]
In theory, zugbruecke should be able to run on any Python installation. I've been trying to run it in Blender, using Blender Python. I installed zugbruecke to user, and when running pip in bpy, Blender created pointers to wenv and zugbruecke, and those packages can now be accessed from inside blender. However, running `import zugbruecke.ctypes` causes a TimeOut error:

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/butter/.local/lib/python3.10/site-packages/zugbruecke/ctypes/__init__.py", line 44, in <module>
    _session = _CtypesSession()
  File "/Users/butter/.local/lib/python3.10/site-packages/zugbruecke/core/session.py", line 129, in __init__
    self._current_session = SessionClient(config=Config(**kwargs))
  File "/Users/butter/.local/lib/python3.10/site-packages/zugbruecke/core/session_client.py", line 116, in __init__
    self._wait_for_server_status_change(target_status=True)
  File "/Users/butter/.local/lib/python3.10/site-packages/zugbruecke/core/session_client.py", line 328, in _wait_for_server_status_change
    raise TimeoutError("session server did not appear")
TimeoutError: session server did not appear
```

When setting up zugbruecke to work outside of Blender,  I encountered this error when something caused Wine to not run Wine Python, such as `arch` being set to `win32` by default instead of `win64`. Having fixed errors like those with configuration files, zugbruecke then worked for my main installation of Python, on homebrew. Using it in bpy still throws this error though, and I can't see why, as `env.ensure()` passes in `session_client.py` with no errors.

TLDR: Expected normal behavior in bpy, got error instead. Running Apple Silicon with Rosetta 2, wine-stable-9.0, zugbruecke 0.2.1, wenv .0.5.1, with bpy running CPython 3.10.12. No logs seem to be generated.

`cat .zugbruecke.json`

````
{
    "arch":"win64",
    "log_write": true,
    "log_level": 100
}
````

`cat .wenv.json`

```
{
    "arch" : "win64",
    "offline" : "true",
    "cache" : "/Users/butter/.local/share/wenv/cache"
}
```

Easiest way to reproduce: 

install Python through homebrew, install Blender for MacOS, use pip to install zugbruecke, `brew install wine-stable`, `cd /Applications/Blender.app/Contents/Resources/3.6/python/bin`, `./python3.10 -m ensurepip`, create config jsons like above in Home (cache was in homebrew because of python installation), `./python3.10 -m pip install zugbruecke` (this should print `Defaulting to user installation because normal site-packages is not writeable`), then run:

`% ./python3.10`
`import zugbruecke.ctypes`

`TimeOut` will give title error, `KeyboardInterrupt` will throw `AttributeError` for `_terminate_on_server`. On the other hand, just running `python` or `python3` at home directory will allow successful import of `zugbruecke.ctypes`.

**Comment by s-m-e** [comment]
I changed the log level structure recently, to match those of Python logging module - try `log_level` set to `logging.DEBUG` i.e. `10` instead of `100`. If I forgot to update the docs in some place, please let me know were.

My *guess* is that you're seeing a difference between x86 Python on Rosetta and native Python on ARM (Apple Silicon). Homebrew's Python might be native ARM while Blender and its Python might be entirely x86 sitting on top of Rosetta. I am guessing that `wenv` somehow fails to handle this case correctly, i.e. fails to fetch a matching version of Python for Windows if running through Rosetta in this case. I'd try to investigate and compare what `wenv` does inside and outside of Blender, specifically the state of the `pythonversion` object around [here](https://github.com/pleiszenburg/wenv/blob/5e86ccaac6dd14f6c2c07c7eed65e89517662cc8/src/wenv/_core/env.py#L338).

Sry, I do not have a Apple Silicon machine to test this hypothesis right now.

**Comment by EclipsedButter** [comment]
Ok, got log:

```
{"level": 20, "platform": "UNIX", "id": "d7ed9b72", "time": 1724756817.72, "pipe": "out", "cnt": "[interpreter] Started with PID 51686."}
{"level": 20, "platform": "UNIX", "id": "d7ed9b72", "time": 1724756817.72, "pipe": "out", "cnt": "[interpreter] Starting stream reader threads ..."}
{"level": 20, "platform": "UNIX", "id": "d7ed9b72", "time": 1724756817.72, "pipe": "out", "cnt": "[interpreter] Stream reader threads started."}
{"level": 20, "platform": "UNIX", "id": "d7ed9b72", "time": 1724756817.72, "pipe": "out", "cnt": "[interpreter] Starting log processing thread ..."}
{"level": 20, "platform": "UNIX", "id": "d7ed9b72", "time": 1724756817.72, "pipe": "out", "cnt": "[interpreter] Log processing thread started."}
{"level": 20, "platform": "UNIX", "id": "d7ed9b72", "time": 1724756817.72, "pipe": "out", "cnt": "[interpreter] STARTED."}
{"level": 20, "platform": "UNIX", "id": "d7ed9b72", "time": 1724756817.72, "pipe": "out", "cnt": "[session-client] Waiting for session-server to be up ..."}
{"level": 40, "platform": "UNIX", "id": "d7ed9b72", "time": 1724756817.82, "pipe": "err", "cnt": "[P] There was no command passed."}
{"level": 20, "platform": "UNIX", "id": "d7ed9b72", "time": 1724756817.82, "pipe": "out", "cnt": "[interpreter] Joining worker threads and queues ..."}
{"level": 20, "platform": "UNIX", "id": "d7ed9b72", "time": 1724756817.82, "pipe": "out", "cnt": "[interpreter] ... worker threads and queues joined."}
{"level": 20, "platform": "UNIX", "id": "d7ed9b72", "time": 1724756847.72, "pipe": "out", "cnt": "[session-client] ... wait timed out (after 30.00 seconds)."}
```

It might be worth testing even on an Intel Mac VM or the like. Blender has an intel version, and it would still be useful to see if this issue is a problem with using Apple Silicon/Rosetta or Blender itself. Blender has some pretty strict file permissions on its stuff, such that I had to have it write the log file to somewhere else instead of the current directory. Running the script in Blender produces a slightly different log:

```
{"level": 20, "platform": "UNIX", "id": "31d963eb", "time": 1724757593.44, "pipe": "out", "cnt": "[interpreter] Started with PID 51745."}
{"level": 20, "platform": "UNIX", "id": "31d963eb", "time": 1724757593.44, "pipe": "out", "cnt": "[interpreter] Starting stream reader threads ..."}
{"level": 20, "platform": "UNIX", "id": "31d963eb", "time": 1724757593.44, "pipe": "out", "cnt": "[interpreter] Stream reader threads started."}
{"level": 20, "platform": "UNIX", "id": "31d963eb", "time": 1724757593.44, "pipe": "out", "cnt": "[interpreter] Starting log processing thread ..."}
{"level": 20, "platform": "UNIX", "id": "31d963eb", "time": 1724757593.44, "pipe": "out", "cnt": "[interpreter] Log processing thread started."}
{"level": 20, "platform": "UNIX", "id": "31d963eb", "time": 1724757593.44, "pipe": "out", "cnt": "[interpreter] STARTED."}
{"level": 20, "platform": "UNIX", "id": "31d963eb", "time": 1724757593.44, "pipe": "out", "cnt": "[session-client] Waiting for session-server to be up ..."}
{"level": 40, "platform": "UNIX", "id": "31d963eb", "time": 1724757593.55, "pipe": "err", "cnt": "[P] python: wenv: command not found"}
{"level": 20, "platform": "UNIX", "id": "31d963eb", "time": 1724757593.55, "pipe": "out", "cnt": "[interpreter] Joining worker threads and queues ..."}
{"level": 20, "platform": "UNIX", "id": "31d963eb", "time": 1724757593.55, "pipe": "out", "cnt": "[interpreter] ... worker threads and queues joined."}
{"level": 20, "platform": "UNIX", "id": "31d963eb", "time": 1724757623.45, "pipe": "out", "cnt": "[session-client] ... wait timed out (after 30.01 seconds)."}
```

Either way, it looks like something happens when `wenv.Env.ensure` is called. In the first log, running `wenv.Env.setup_pip` prints `There was no command passed.` This is a behavior I incidentally introduced by forcing setup_pip to run its subprocesses (when in offline mode) with `shell=True` on `FileNotFoundException`, as Blender could not otherwise access the `wenv` command. Note that when running in Blender however, the same spot in the log prints `wenv: command not found`. That said, they both demonstrate the same end result. I will try tracking down where `wenv` is being called there and see if resolving that fixes things.

**Comment by s-m-e** [comment]
If it's simply write permissions, the location of the [wine prefix](https://github.com/pleiszenburg/wenv/blob/5e86ccaac6dd14f6c2c07c7eed65e89517662cc8/src/wenv/_core/config.py#L135) might be your limitation. You can set it as an option in `wenv` to a writeable location, otherwise you'll end up somewhere in Blender's directory structure. The most common use case I see with `wenv` etc is Python virtual environments, which are usually user writeable. With software that embeds Python itself, this might be different. However, you should see something in the logs ... if not, this needs fixing on my part.

**Comment by EclipsedButter** [comment]
Welp, it doesn't seem like the location of the wineprefix matters. It already points to `~/.local`, which is writeable (I've been running it with `"offline" : "true"` because otherwise Blender certifi blocks the request). I tested several different configurations, and even putting the wenv and zugbruecke installations in Blender's python directory doesn't change anything, as it still has the exact same session error. Additionally, the `There was no command passed.` and `wenv: command not found` statements appear completely benign, and if they are going to cause errors, I assume they will only happen after this `TimeOutError` is resolved. 

I also checked, and Rosetta isn't causing problems here, as both Blender and its Python are `Mach-O 64-bit executable arm64`. Can we know if this also happens on Intel Macs? If so, it could indicate a problem with initializing Wine Python from custom Python installations.

**Comment by s-m-e** [comment]
When you're dealing with zugbruecke/wenv, there are two Python interpreters: One on the MacOS side and one on the Wine side, a Windows build of Python. The two Pythons should by default use one single architecture - so if it's an ARM variant on the MacOS side, it should be an ARM variant on the Wine side as well. This could be an issue, if say Wine runs on Rosetta and zugbruecke throws an ARM variant at it (or the other way round) Theoretically, zugbruecke/wenv should work across different architectures, i.e. an ARM variant on MacOS and an x86 variant on Wine (or again the other way round) - I just never anticipated this case hence it is not something that can be easily configured yet. It's a bit of a rabbit hole: What's the architecture of Wine (i.e. native ARM or Rosetta) and, independently, the Windows build of Python that gets pulled (via `ensure`)?

**Comment by EclipsedButter** [comment]
Hmm. Well, I downloaded Blender macOS Intel for the heck of it to see what would happen, since its python is `Mach-O 64-bit executable x86_64`. While at the end of the day I got the exact same error (concerning), the logs are a bit more descriptive this time:

```
{"level": 20, "platform": "UNIX", "id": "8dbbc7e4", "time": 1724769889.85, "pipe": "out", "cnt": "[interpreter] Started with PID 53815."}
{"level": 20, "platform": "UNIX", "id": "8dbbc7e4", "time": 1724769889.85, "pipe": "out", "cnt": "[interpreter] Starting stream reader threads ..."}
{"level": 20, "platform": "UNIX", "id": "8dbbc7e4", "time": 1724769889.85, "pipe": "out", "cnt": "[interpreter] Stream reader threads started."}
{"level": 20, "platform": "UNIX", "id": "8dbbc7e4", "time": 1724769889.85, "pipe": "out", "cnt": "[interpreter] Starting log processing thread ..."}
{"level": 20, "platform": "UNIX", "id": "8dbbc7e4", "time": 1724769889.85, "pipe": "out", "cnt": "[interpreter] Log processing thread started."}
{"level": 20, "platform": "UNIX", "id": "8dbbc7e4", "time": 1724769889.85, "pipe": "out", "cnt": "[interpreter] STARTED."}
{"level": 20, "platform": "UNIX", "id": "8dbbc7e4", "time": 1724769889.85, "pipe": "out", "cnt": "[session-client] Waiting for session-server to be up ..."}
{"level": 20, "platform": "WINE", "id": "8dbbc7e4", "time": 1724769890.42, "pipe": "out", "cnt": "[session-server] STARTING ..."}
{"level": 40, "platform": "UNIX", "id": "8dbbc7e4", "time": 1724769890.47, "pipe": "err", "cnt": "[P] Traceback (most recent call last):\r"}
{"level": 40, "platform": "UNIX", "id": "8dbbc7e4", "time": 1724769890.47, "pipe": "err", "cnt": "[P]   File \"D:\\obj\\windows-release\\37amd64_Release\\msi_python\\zip_amd64\\runpy.py\", line 193, in _run_module_as_main\r"}
{"level": 40, "platform": "UNIX", "id": "8dbbc7e4", "time": 1724769890.47, "pipe": "err", "cnt": "[P]   File \"D:\\obj\\windows-release\\37amd64_Release\\msi_python\\zip_amd64\\runpy.py\", line 85, in _run_code\r"}
{"level": 40, "platform": "UNIX", "id": "8dbbc7e4", "time": 1724769890.47, "pipe": "err", "cnt": "[P]   File \"C:\\python-3.7.4.stable\\lib\\site-packages\\zugbruecke\\_server_.py\", line 81, in <module>\r"}
{"level": 40, "platform": "UNIX", "id": "8dbbc7e4", "time": 1724769890.47, "pipe": "err", "cnt": "[P]     run()\r"}
{"level": 40, "platform": "UNIX", "id": "8dbbc7e4", "time": 1724769890.47, "pipe": "err", "cnt": "[P]   File \"C:\\python-3.7.4.stable\\lib\\site-packages\\zugbruecke\\_server_.py\", line 72, in run\r"}
{"level": 40, "platform": "UNIX", "id": "8dbbc7e4", "time": 1724769890.47, "pipe": "err", "cnt": "[P]     _ = SessionServer(Config(**parameter))\r"}
{"level": 40, "platform": "UNIX", "id": "8dbbc7e4", "time": 1724769890.47, "pipe": "err", "cnt": "[P]   File \"C:\\python-3.7.4.stable\\lib\\site-packages\\zugbruecke\\core\\session_server.py\", line 84, in __init__\r"}
{"level": 40, "platform": "UNIX", "id": "8dbbc7e4", "time": 1724769890.47, "pipe": "err", "cnt": "[P]     self._log.info(\"[session-server] STARTING ...\")\r"}
{"level": 40, "platform": "UNIX", "id": "8dbbc7e4", "time": 1724769890.47, "pipe": "err", "cnt": "[P]   File \"C:\\python-3.7.4.stable\\lib\\site-packages\\zugbruecke\\core\\log.py\", line 103, in info\r"}
{"level": 40, "platform": "UNIX", "id": "8dbbc7e4", "time": 1724769890.47, "pipe": "err", "cnt": "[P]     self._process_raw(*raw_messages, pipe=\"out\", level=INFO)\r"}
{"level": 40, "platform": "UNIX", "id": "8dbbc7e4", "time": 1724769890.47, "pipe": "err", "cnt": "[P]   File \"C:\\python-3.7.4.stable\\lib\\site-packages\\zugbruecke\\core\\log.py\", line 135, in _process_raw\r"}
{"level": 40, "platform": "UNIX", "id": "8dbbc7e4", "time": 1724769890.47, "pipe": "err", "cnt": "[P]     self._process(message)\r"}
{"level": 40, "platform": "UNIX", "id": "8dbbc7e4", "time": 1724769890.47, "pipe": "err", "cnt": "[P]   File \"C:\\python-3.7.4.stable\\lib\\site-packages\\zugbruecke\\core\\log.py\", line 146, in _process\r"}
{"level": 40, "platform": "UNIX", "id": "8dbbc7e4", "time": 1724769890.47, "pipe": "err", "cnt": "[P]     message.store(self._f)\r"}
{"level": 40, "platform": "UNIX", "id": "8dbbc7e4", "time": 1724769890.47, "pipe": "err", "cnt": "[P]   File \"C:\\python-3.7.4.stable\\lib\\site-packages\\zugbruecke\\core\\log.py\", line 235, in store\r"}
{"level": 40, "platform": "UNIX", "id": "8dbbc7e4", "time": 1724769890.47, "pipe": "err", "cnt": "[P]     with open(os.path.join(os.getenv('HOME'),f'zugbrueckeLog/{self._id}.txt'), mode=\"a+\", encoding=\"utf-8\") as f:\r"}
{"level": 40, "platform": "UNIX", "id": "8dbbc7e4", "time": 1724769890.47, "pipe": "err", "cnt": "[P]   File \"D:\\obj\\windows-release\\37amd64_Release\\msi_python\\zip_amd64\\ntpath.py\", line 76, in join\r"}
{"level": 40, "platform": "UNIX", "id": "8dbbc7e4", "time": 1724769890.47, "pipe": "err", "cnt": "[P] TypeError: expected str, bytes or os.PathLike object, not NoneType\r"}
{"level": 20, "platform": "UNIX", "id": "8dbbc7e4", "time": 1724769890.47, "pipe": "out", "cnt": "[interpreter] Joining worker threads and queues ..."}
{"level": 20, "platform": "UNIX", "id": "8dbbc7e4", "time": 1724769890.47, "pipe": "out", "cnt": "[interpreter] ... worker threads and queues joined."}
{"level": 20, "platform": "UNIX", "id": "8dbbc7e4", "time": 1724769919.86, "pipe": "out", "cnt": "[session-client] ... wait timed out (after 30.01 seconds)."}
```

It hit a lot more errors, that might be Rosetta breaking something....

(Ignore anything about `zugbrueckeLog`, that was again me bypassing the permission error on writing the log in the current directory)

**Comment by s-m-e** [comment]
The good news in your error / traceback is that the Windows version of Python running on Wine is actually running. It's causing the error that is captured and forwarded to the Python interpreter running on MacOS. I'd say reset the code to my original and see what happens. Your error is caused by `os.getenv('HOME')` resulting in `None`, likely because `HOME` is not set (again, Wine is its independent thing with its own independent environment variables and I make sure not to forward all environment variables from the host, in your case MacOS, to it for the sake of not making it run even shakier that it is on its own).

**Comment by EclipsedButter** [comment]
You are right! The only two changes I made to your code were: to `wenv/_core/env.py`, where I put in an exception for `Env.setup_pip` to use the shell when `wenv` command was not recognized--a fix specifically for when running Blender, as it doesn't have access to it otherwise, whereas running Blender's `python3.10` in the terminal does not trigger this exception; and a change to `zugbruecke/core/log.py`, which is where `os.getenv` is being called. With logging on and at 10, I was getting a permission error when the logger tried to write the log file to the current directory, so I excepted that error and had it also write to a log file pointed to by `$HOME`. You are right that it is causing the error and turning off logging solves it. It only occurs when running Blender's Python3.11 on Rosetta, so it won't happen if running the ARM version of Blender, which is still experiencing the session error....

I guessed that part is not something to be fixed so easily. Intel Blender is on a later version anyway (even though it runs slower), so I can work with that. That said, it is still interesting to me that I am able to run zugbruecke just fine on my main python3.10 installation.... which is arm64.

**Comment by EclipsedButter** [comment]
Well that was rather dense of me, thinking it would just work like that. With your changes, importing `zugbruecke.ctypes` works in the Blender Python shell in terminal, but not within Blender (x64) itself.

Traceback:
```
[1;30m(1724797264.92/31bb2d72[0;0m)[1;34mUNIX [0;0m[1;31me[0;0m: [1;37m[P] python: wenv: command not found[0;0m
Traceback (most recent call last):
  File "<blender_console>", line 1, in <module>
  File "/Applications/Blender 2.app/Contents/Resources/4.2/python/lib/python3.11/site-packages/zugbruecke/ctypes/__init__.py", line 44, in <module>
    _session = _CtypesSession()
               ^^^^^^^^^^^^^^^^
  File "/Applications/Blender 2.app/Contents/Resources/4.2/python/lib/python3.11/site-packages/zugbruecke/core/session.py", line 129, in __init__
    self._current_session = SessionClient(config=Config(**kwargs))
                            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/Applications/Blender 2.app/Contents/Resources/4.2/python/lib/python3.11/site-packages/zugbruecke/core/session_client.py", line 116, in __init__
    self._wait_for_server_status_change(target_status=True)
  File "/Applications/Blender 2.app/Contents/Resources/4.2/python/lib/python3.11/site-packages/zugbruecke/core/session_client.py", line 328, in _wait_for_server_status_change
    raise TimeoutError("session server did not appear")
TimeoutError: session server did not appear
```

Log:
```
{"level": 20, "platform": "UNIX", "id": "31bb2d72", "time": 1724797264.81, "pipe": "out", "cnt": "[interpreter] Started with PID 56783."}
{"level": 20, "platform": "UNIX", "id": "31bb2d72", "time": 1724797264.81, "pipe": "out", "cnt": "[interpreter] Starting stream reader threads ..."}
{"level": 20, "platform": "UNIX", "id": "31bb2d72", "time": 1724797264.81, "pipe": "out", "cnt": "[interpreter] Stream reader threads started."}
{"level": 20, "platform": "UNIX", "id": "31bb2d72", "time": 1724797264.81, "pipe": "out", "cnt": "[interpreter] Starting log processing thread ..."}
{"level": 20, "platform": "UNIX", "id": "31bb2d72", "time": 1724797264.81, "pipe": "out", "cnt": "[interpreter] Log processing thread started."}
{"level": 20, "platform": "UNIX", "id": "31bb2d72", "time": 1724797264.81, "pipe": "out", "cnt": "[interpreter] STARTED."}
{"level": 20, "platform": "UNIX", "id": "31bb2d72", "time": 1724797264.81, "pipe": "out", "cnt": "[session-client] Waiting for session-server to be up ..."}
{"level": 40, "platform": "UNIX", "id": "31bb2d72", "time": 1724797264.92, "pipe": "err", "cnt": "[P] python: wenv: command not found"}
{"level": 20, "platform": "UNIX", "id": "31bb2d72", "time": 1724797264.92, "pipe": "out", "cnt": "[interpreter] Joining worker threads and queues ..."}
{"level": 20, "platform": "UNIX", "id": "31bb2d72", "time": 1724797264.92, "pipe": "out", "cnt": "[interpreter] ... worker threads and queues joined."}
{"level": 20, "platform": "UNIX", "id": "31bb2d72", "time": 1724797294.81, "pipe": "out", "cnt": "[session-client] ... wait timed out (after 30.00 seconds)."}
```

There's a fairly obvious `wenv: command not found` in there, but I'm not sure how that's happening such that it only triggers a critical error instead of raising an exception, when failing to start `wenv` should probably do that. Furthermore, because it doesn't raise an exception and just has `[P]` instead of a filename, I can't locate where it is.

**Comment by s-m-e** [comment]
> works in the Blender Python shell in terminal, but not within Blender (x64) itself

There is a clue together with your error (that is not caught, sry). `wenv` should be in `PATH` but it is not from within Blender. Blender likely does some cleaning on `PATH` within. It's easy to confirm. Both from outside and from within Blender, check if the location of `wenv` is in `PATH`, simply by checking the contents of ~~`os.getenv('HOME')`~~. You can carefully try to add it to `PATH` before importing `zugbruecke` but be aware that the Blender folks clean `PATH` for a number of potential issues.

**Comment by s-m-e** [comment]
The contents of `os.getenv('PATH')`, not HOME ... sry.

**Comment by EclipsedButter** [comment]
Once again, you knew exactly what was going on. I added `:/opt/homebrew/bin` to `os.environ['PATH']` and it started working. Interestingly, this only worked exclusively for the brew installation; I have a bunch of other `bin` directories--even one in Blender.app, in the same `bin` as `python3.11`--but none of those worked, I'll check the logs for that. Overall, a bunch of hacky solutions and it works, but I'll try to optimize and clean up this approach. Now with `ctypes.WinDLL`, I only have to run `os.add_dll_directory`... oh. oh no....

<sub>[Availability](https://docs.python.org/3/library/os.html#os.add_dll_directory): Windows.</sub>




well that's for another day. Thank you for all of your help, it was super timely and effective!

**Comment by EclipsedButter** [comment]
Alright, some `os.environ` to substitute for `add_dll_directory` and to put `homebrew/bin` into `PATH` and it all works. In fact, undoing most of my changes, it also runs on Blender ARM64. Thanks again!

**Comment by s-m-e** [comment]
![zugbruecke_demo](https://github.com/user-attachments/assets/71a0e91a-bbb7-4248-addf-0be2c40e7ec7)

This is not the first time `zugbruecke` was used in Blender, though the last time I (successfully) attempted it was on Linux a few years ago (2018). Besides, this is the first (at least for me) confirmed use on an ARM-based MacOS.

I'd be really happy about PRs documenting/summarizing how you ultimately did it and - potentially - what needs to change in `zugbruecke` to make this usecase easier.

**Comment by EclipsedButter** [comment]
Well I had made a fork and was going to do a PR showing all the changes, but I went back through and noticed that none of my changes were necessary... so no PR  :(

The only error that ever happens is when logging is on, because all of the contents of Blender.app aren't writable by a non-user, but that's true for more than just Blender. I can document what I think is the simplest way to set it up though:
Open Blender
Go to scripts
In the console, type the following
```
import pip
pip.main(['install', 'zugbruecke'])
pip.main(['install', 'wenv'])
import os
os.environ['PATH'] += ':/opt/homebrew/bin'
```
That's it. `import zugbruecke.ctypes` can be run immediately after. A couple things were tripping me up before that caused all the issues I had:

For one, I had assumed that running Blender's python executable (`/Applications/Blender.app/Contents/Resources/4.2/python/bin/python3.11`) was equivalent to running python in Blender's scripting tab. For the most part, it is, except for what you pointed out: Blender cleans `PATH` on launch. So I would run `./python3.11 -m pip install zugbruecke` in terminal, and because `PATH` hadn't been cleaned, it would notice that `pip` had already installed `zugbruecke` to homebrew and `.local`, and would just return `requirement already satisfied`. But upon launching Blender, those two directories will no longer be available in `PATH`, so Blender won't find `zugbruecke` or `wenv`. Furthermore, it won't find Wine anywhere either, unless the Wine executable was installed directly to `/usr/bin`, and Wine is usually installed through homebrew.

Additionally, things wouldn't work with the default architecture of `win32`. Creating a `.zugbruecke.json` and `.wenv.json` file and doing `arch : win64` for both of them, as per your docs, fixes that issue. I also like to add `offline: true` in the wenv config, since certifi blocked things a couple times inside Blender. I had thought this would also require running `wenv init` and `wenv cache` somewhere, but I'm not seeing a cache anywhere and it's still running fine...

If logging is on, you might have to edit `zugbruecke/core/log.py/Message.store` to write to a file identified by self._id and outside of the Blender.app directory (that's what I did at least) to avoid permission errors.

If importing a python add-on that uses zugbruecke, you would, after running the `pip` commands in scripting, edit the addon to add the homebrew bin to `PATH` in the addon's `__init__.py` (or whatever file runs on addon activation) before activating the addon.


I think that's about it. There would of course be differences in setup if pip and Wine aren't installed to brew, or in different places entirely. Blender should come pre-installed with pip anyway. All of this said, I can't help but think there is probably a neater way to do this; like you said, Blender probably cleans `PATH` for good reason, so a way for the package locate the wenv and Wine executables without modifying might be a good idea. At the same time, Blender reduces path to just `/usr/bin` and `/usr/sbin`, which I don't think would often contain anything that conflicts with what would be in hombrew bin, so I don't foresee any issues with it.
